### PR TITLE
Implement initial Ordered PubSub Support

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds initial support for ordered pubsub on AWS, GCP and EncoreCloud.

It uses a new `OrderingAttribute` configuration from the topic to extract the ordering key from the message attributes at the time of the publish.

### CODE REVIEW

File: runtime/pubsub/internal/types/public.go [hunk 1]
```diff
@@ -90,10 +90,49 @@ type TopicConfig struct {
...
+	//  topic.Publish(ctx, &UserEvent{UserID: "1", Action: "login"})  // This message will be delivered before the logout
+	//  topic.Publish(ctx, &UserEvent{UserID: "2", Action: "login"})  // This could be delivered at any time because it has a different user id
+	//  topic.Publish(ctx, &UserEvent{UserID: "1", Action: "logout"}) // This message will be delivered after the first message
+	//
+	// By using OrderingAttribute, the throughput will be limited depending on the cloud provider:
+	//
+	// - AWS: 300 messages per second for the topic (see [AWS SQS Quotas]).
+	// - GCP: 1MB/s for each ordering key (see [GCP PubSub Quotas]).
+	//
+	// Note: OrderingAttribute is not supported for local development.
```
Comment (eandre) [inline runtime/pubsub/internal/types/public.go:131]:
```suggestion
	// Note: OrderingAttribute currently has no effect during local development.
```

File: runtime/pubsub/internal/types/public.go [hunk 2]
```diff
@@ -90,10 +90,49 @@ type TopicConfig struct {
...
+	//
+	// [AWS SQS Quotas]: https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/quotas-messages.html
+	// [GCP PubSub Quotas]: https://cloud.google.com/pubsub/quotas#resource_limits
+	//
+	//publicapigen:drop
```
Comment (eandre) [inline runtime/pubsub/internal/types/public.go:136]: Remove, if we're ready to go live with it

File: v2/app/testdata/pubsub_err_ordering_attribute_missing.txt [hunk 3]
```diff
@@ -0,0 +1,42 @@
+# Verify that pub sub is parsed
+! parse
+err 'PubSub message attributes must not be prefixed with "encore".'
```
Comment (eandre) [inline v2/app/testdata/pubsub_err_ordering_attribute_missing.txt:3]: This seems wrong; there is no such attribute

Comment (DomBlack) [reply v2/app/testdata/pubsub_err_ordering_attribute_missing.txt:3]: The v2 parser tests don't actually look at the `err` flag, so I've removed this (it was copy/pasted from a previous test)

File: runtime/pubsub/internal/types/public.go [hunk 4]
```diff
@@ -90,10 +90,49 @@ type TopicConfig struct {
 	// This field is required.
 	DeliveryGuarantee DeliveryGuarantee
 
-	// OrderingKey is the name of the message attribute used to group
```
[the remark this answers is not in the document]
Comment (eandre) [reply runtime/pubsub/internal/types/public.go:93]: It's specifying a schema, so it is specifying what is the ordering key. Just like when you create a table in postgres you're defining a "primary key" (by naming a column), not a "primary column".

[the remark this answers is not in the document]
Comment (DomBlack) [reply runtime/pubsub/internal/types/public.go:93]: To be clear; we're actually not renaming anything here (officially) as we never released / publicised the original API design. 

I get what you're saying around the Primary Key thing, but given we already confused the field once, and in the GCP API there's a field called "OrderingKey" (AWS it's "GroupingID") I felt we wanted to reduce the risk of that gain.

The other thing I prefer about this name is it's clear it's for an Attribute, meaning if we wanted to change it to a FieldName in the future, there's no abimgilotly as we could add a "OrderingField" option for that.

[the remark this answers is not in the document]
Comment (eandre) [reply runtime/pubsub/internal/types/public.go:93]: Yeah let's merge it as-is. Will mull it over

**Pull request merged**